### ISSUE TITLE
crypto/ed25519: Caching public key decompression

### ISSUE THREAD

**Issue by ValarDragon** [tendermint/tendermint#1784]
For a new node syncing to the chain, they have to process many signatures from the same validators. The same applies to full nodes verifying signatures. Under Ed25519, these public keys have to undergo a point decompression. This point decompression can be trivially cached (see https://github.com/tendermint/ed25519/pull/8). Benchmarking on that PR, caching this point decompression saves `10%` of computation time per verify, for all verifications after the first one. (The first verify takes the same exact amount of time)

If we ever implement Ed25519 batch verification, this savings becomes `> 20%` of the batch verify time, since batch verification doesn't do anything for the point decompression employed. 

It has been decided to hold off on integrating this into tendermint, until sync times become a problem for people. For reference, Ed25519 signature verification takes `65%` of the time of a CheckTx + DeliverTx combination, implying that it is significant even when accounting for the account mapper / state reads. 

The way this would likely be implemented is by changing the internal type of Ed25519 to be a struct with the relevant cached values. The alternative is adding a cache method to every public key type -- I'm not opposed to this as it avoids the extra temporary memory cost of most keys, and I think more key types will benefit from this caching paradigm in a way that shouldn't be ran on keys that are used only once, unlike ed25519. The downside of the cache method is that its confusing / API we have to maintain. (Not that confusing imo though). The downsides of switching Ed25519 to a struct is that every Ed25519 public key in memory is going to take an extra 64 bytes. (It will still marshal the same way, and therefore be the exact same in state). Arguably that isn't an issue, since this provides more significant computational savings if the key is ever reused, and if it isn't reused it will be garbage collected almost immediately. However it must be cautioned then whenever public keys are encoded for being put into state, only the marshalled version is encoded. (I believe this is already the case, so there isn't any concern)

In general, this is a relatively easy optimization we can take advantage of, once we want more speed out of our signature verifications.

**Comment by ValarDragon** [comment]
Turns out there is even more caching we can do for free! (Well free computation-wise, costs memory still)

The Ed25519 scheme we're using uses 4-bit windowed computation for unknown bases. (This means that for points that aren't the base point, e.g. `A`, the multiples `[3..15] * A` are cached. (Ignoring the evens, due to those being trivial to compute via inverses). Currently these are computed on every verify for the pubkey. We can cache these also at no cost. The benefits will be less dramatic than point-decompression (I think, haven't benchmarked yet), but it should still be noticeable improvement. I think the memory trade-off for caching these will be worth it, since this is stuff thats computed everytime.

**Comment by robert-zaremba** [comment]
What is the story with this task? It seams that it's an easy optimization. BTW - are we already using batched verification in tendermint?

**Comment by tessr** [comment]
No, we don't yet have batched verification. That (and most of the other requested crypto work) is currently slated for a release shortly after 1.0.

**Comment by hdevalence** [comment]
One approach to caching decompressed public keys is the [`VerificationKey`](https://docs.rs/ed25519-zebra/2.2.0/ed25519_zebra/struct.VerificationKey.html) structure in `ed25519-zebra`.

While it's true that lookup tables can also be cached, this is further along the "engineering effort"/"performance benefit" tradeoff, since without support from the underlying curve library (e.g., a structure [like this](https://docs.rs/curve25519-dalek/3.0.0/curve25519_dalek/edwards/struct.VartimeEdwardsPrecomputation.html)) it involves breaking through abstraction layers.  And, once precomputation is acceptable, there are more tradeoffs to consider, like:

* rather than using a small lookup table, should a large lookup table be used? (the existing size of the lookup table is tuned on the assumption that it's only used once, so it balances construction time vs time savings on a one-off basis)
* is single verification actually the relevant problem? (is it possible to use batch verification instead, which provides much greater savings)?
* how many public keys are in use, and how frequently are they used? (since precomputation is a kind of memory-time tradeoff, if there are many public keys, the required memory may be larger than expected, and memory tradeoffs are discontinuous because of cache size effects).

If Tendermint is planning to use `ed25519consensus`, the right place to put this kind of precomputation is probably there.

**Comment by KaycoinsInc** [comment]
That is a nice feature. Please explain simply in easy to understand terms please because i do not know all the verbage yet.Just when I use this instead of signing with my applications or git. Sorry if dumb question.

**Comment by KaycoinsInc** [comment]
This keeps those more secure by them not being exposed correct so use this more.

**Comment by robert-zaremba** [comment]
> how many public keys are in use, and how frequently are they used? 

@hdevalence cosmos chains have 100-300 validators. Gaia (the Cosmos Hub) has around 150. So I would say there are 150 keys, used in every round (block). They don't change that frequently.

What do you mean with `ed25519consensus`?

**Comment by marbar3778** [comment]
Single key verification has increased by 30-40% (on master). Once I complete my PR on batch verification it will be at least 40%, if not double. I would categorize this issue as a low priority, especially since we don't have anyone internally to work on this.

**Comment by robert-zaremba** [comment]
> Single key verification has increased by 30-40% (on master)

There was some recent change which increased the verification time by 30-40%? What was that?

**Comment by marbar3778** [comment]
https://github.com/tendermint/tendermint/pull/5760

The dependency got updated, we just updated the dep in our repo.

**Comment by tessr** [comment]
Just to be clear: The time _decreased_ as a result of updating the dependency. This is a good thing!

**Comment by robert-zaremba** [comment]
> Just to be clear: The time _decreased_ as a result of updating the dependency. This is a good thing!

Thanks, I read the OP that the verification increased and assumed it's about time verification :sweat_smile:

**Comment by marbar3778** [comment]
completed via #6526
